rechunker/api.py

```diff
@@ -10,12 +10,100 @@
+    return tuple(shape)
+
+
+def _get_dims_from_zarr_array(z_array):
+    # use Xarray convention
+    # http://xarray.pydata.org/en/stable/internals.html#zarr-encoding-specification
+    return z_array.attrs["_ARRAY_DIMENSIONS"]
+
+
+def _zarr_empty(shape, store_or_group, chunks, dtype, name=None):
+    # wrapper that maybe creates the array within a group
+    if name is not None:
+        assert isinstance(store_or_group, zarr.hierarchy.Group)
+        return store_or_group.empty(name, shape=shape, chunks=chunks, dtype=dtype)
+    else:
+        return zarr.empty(shape, chunks=chunks, dtype=dtype, store=store_or_group)
+
+
+def rechunk(
+    source, target_chunks, max_mem, target_store, temp_store=None,
+):
+    """
+    Rechunk a Zarr Array or Group
+
+    Parameters
+    ----------
+    source : zarr.Array or zarr.Group
+    target_chunks : tuple or dict
+        The desired chunks of the array after rechunking.
+    max_mem : int
+        The amount of memory (in bytes) that workers are allowed to use
+    target_store : str, MutableMapping, or zarr.Store object
+        The location in which to store the final, rechunked result
+    temp_store : str, MutableMapping, or zarr.Store object, optional
+        Location of temporary store for intermediate data. Can be deleted
+        once rechunking is complete.
+    """
+
+    # these options are not tested yet; don't include in public API
+    kwargs = dict(
+        source_storage_options={}, temp_storage_options={}, target_storage_options={},
+    )
+
+    if isinstance(source, zarr.hierarchy.Group):
+        if not isinstance(target_chunks, dict):
+            raise ValueError(
+                "You must specificy ``target-chunks`` as a dict when rechunking a group."
+            )
+
+        stores_delayed = []
+
+        if temp_store:
+            temp_group = zarr.group(temp_store)
+        target_group = zarr.group(target_store)
+        target_group.attrs.update(source.attrs)
+
+        for array_name, array_target_chunks in target_chunks.items():
+            delayed = _rechunk_array(
+                source[array_name],
+                array_target_chunks,
+                max_mem,
+                target_group,
+                temp_store_or_group=temp_group,
+                name=array_name,
+                **kwargs,
+            )
+            stores_delayed.append(delayed)
+
+        return stores_delayed
+
+    elif isinstance(source, zarr.core.Array):
+        return _rechunk_array(
+            source,
+            target_chunks,
+            max_mem,
+            target_store,
+            temp_store_or_group=temp_store,
+            **kwargs,
+        )
+
+    else:
+        raise ValueError("Source must be a Zarr Array or Group.")
+
+
+def _rechunk_array(
     source_array,
     target_chunks,
     max_mem,
-    target_store,
-    temp_store=None,
+    target_store_or_group,
+    temp_store_or_group=None,
+    name=None,
     source_storage_options={},
```

Comment:
Nitpick, mutable default values can cause surprises if the dictionaries are mutated. Better to use None and create the empty dict internally.